frame/staking/src/pallet/impls.rs

```diff
@@ -755,18 +755,32 @@ impl<T: Config> Pallet<T> {
 	/// Get the targets for an upcoming npos election.
 	///
 	/// This function is self-weighing as [`DispatchClass::Mandatory`].
-	pub fn get_npos_targets() -> Vec<T::AccountId> {
-		let mut validator_count = 0u32;
-		let targets = Validators::<T>::iter()
-			.map(|(v, _)| {
-				validator_count.saturating_inc();
-				v
-			})
-			.collect::<Vec<_>>();
+	pub fn get_npos_targets(maybe_max_len: Option<usize>) -> Vec<T::AccountId> {
+		let max_allowed_len = maybe_max_len.unwrap_or_else(|| T::TargetList::count() as usize);
+		let mut all_targets = Vec::<T::AccountId>::with_capacity(max_allowed_len);
+		let mut targets_seen = 0;
+
+		let mut targets_iter = T::TargetList::iter();
+		while all_targets.len() < max_allowed_len &&
+			targets_seen < (NPOS_MAX_ITERATIONS_COEFFICIENT * max_allowed_len as u32)
+		{
+			let target = match targets_iter.next() {
```

Comment:
TODO for me or someone else to test the new snapshot creation logic on Polkadot+Kusama in WASM before this is merged.

Reply:
successfully executed the snapshot creation of Polkadot using this branch's code:

```

2022-09-14 07:47:21.897  INFO                 main runtime::frame-support: ⚠️ XcmPallet declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `StorageVersion(0)`
2022-09-14 07:47:21.909  INFO                 main runtime::staking: [12028824] 💸 generated 994 npos targets
2022-09-14 07:47:22.646  INFO                 main runtime::staking: [12028824] 💸 generated 22500 npos voters, 199 from validators and 22301 nominators
2022-09-14 07:47:22.646  INFO                 main runtime::election-provider: [#12028824] 🗳  creating a snapshot with metadata SolutionOrSnapshotSize { voters: 22500, targets: 994 }
2022-09-14 07:47:22.646 DEBUG                 main runtime::election-provider: [#12028824] 🗳  snapshot pre-calculated size 8440202
2022-09-14 07:47:22.652  INFO                 main runtime::election-provider: [#12028824] 🗳  Starting signed phase round 537.
2022-09-14 07:47:22.652 DEBUG                 main runtime::inclusion-inherent: [enter_inner] parent_header=0xb60f61c5b845d61a04bc2c9d86beb0042c6f1c5c66961a9756f729ebd8ad4fe0 bitfields.len(): 196, backed_candidates.len(): 15, disputes.len(): 0
2022-09-14 07:47:22.667 DEBUG                 main runtime::system: [12028824] 0 extrinsics, length: 27739 (normal 0%, op: 0%, mandatory 0%) / normal weight:Weight(ref_time: 0) (0%) op weight Weight(ref_time: 0) (0%) / mandatory weight Weight(ref_time: 3285000008000) (0%)
```

Command used: 

```

RUST_LOG=remote-ext=debug,runtime=debug cargo run --release --features=try-runtime try-runtime --execution=wasm --chain=polkadot-dev --no-spec-name-check execute-block --overwrite-wasm-code live --uri ws://localhost:9999 --at 0x410fd68f263d261e345d15124f5357f265431e2462113726afed2f3620029e72 -s snap0x41

OR 

RUST_LOG=remote-ext=debug,runtime=debug cargo run --release --features=try-runtime try-runtime --execution=wasm --chain=polkadot-dev execute-block --no-spec-name-check --block-ws-uri ws://localhost:9999 --block-at 0x410fd68f263d261e345d15124f5357f265431e2462113726afed2f3620029e72  --overwrite-wasm-code snap -s snap0x41
```

https://polkadot.subscan.io/block/12028824

is chosen because the `UnsignedPhaseStarted` event, which signals snapshot creation, is triggered in this block.